Expand "Local development"

tbeseda opened the pull request:
> 🚧   **WIP**: creating draft PR to coordinate with Ryan, but all feedback welcome :)

## Changes

### Expand "_Local development_"

I added an example for "_Creating a new resource_". I used a hypothetical "hit counter" example. `arc init` is used for scaffolding (mentioned in [this issue](https://github.com/architect/arc.codes/issues/312))
This type of workflow was a point of interest as I was building my first app.

Stub out "_Debugging_" section that may or may not be necessary. Leads into testing.

### Move "_Testing_" to its own document

Testing is definitely a part of "local dev" but probably warrants its own doc.

## More ideas for "Local development"

* provide a short implementation of the "hit counter" event feature, just to tie it all together
* use inline tabs to demonstrate Ruby and Python alongside the Node example
* elaborate on what Architect Sandbox emulates locally (and how?)
  * ie. queues, events, dynalite
  * this is covered in `reference/cli/sandbox`, but there might be an opportunity to tout some of the capabilities without requiring the reader to dig too far.

~I avoided second person, but after [research](https://docs.microsoft.com/en-us/style-guide/grammar/person) on [style guides](https://developers.google.com/style/person) by [large orgs](https://developer.salesforce.com/docs/atlas.en-us.salesforce_pubs_style_guide.meta/salesforce_pubs_style_guide/style_second_person.htm), I'll change it.~

---

- [x] Forked the repo and created your branch from `master`
- [x] Made sure tests pass (run `npm it` from the repo root)
- [x] Updated relevant documentation internal to this repo (e.g. `readme.md`, help docs, inline docs & comments, etc.)
- [x] Linked to any related issues, PRs, etc. below that may relate to, consume, or necessitate these changes

ryanblock commented on src/http/any-catchall/index.js:
```diff
@@ -1,14 +1,15 @@
 
 // middleware proxy s3 assets
-let asap = http.proxy({
+let static = asap({
```
Fyi, careful with `static`, it's reserved! https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Lexical_grammar#future_reserved_keywords

ryanblock commented on src/views/docs/en/guides/developer-experience/local-development.md:
```diff
@@ -4,112 +4,103 @@ category: Developer experience
+│           ├── index.js
+│           └── config.arc
+│
+└── app.arc
```
Hmm, some funky unicode artifacts here?
![image](https://user-images.githubusercontent.com/200964/128097791-4d618aaa-d70f-4144-8034-6a156d46b928.png)

tbeseda replied:
Super weird. that line is the same as the other `└── app.arc` instances in project-layout.md and queues.md.  
and the `─` character that shows up garbled in your screenshot is used in each of the lines above the app.arc line in your screenshot.

The issue must be between the code and the HTML output, maybe the Markdown compiler. but I'm having a hard time reproducing that garbled text.  
I'm already taking a look there since there seems to be an issue with my highlighter.js change, so I can take some time to look more at this.

ryanblock commented on src/views/docs/en/guides/developer-experience/local-development.md:
```diff
@@ -4,112 +4,103 @@ category: Developer experience
-  await sandbox.end()
-  t.ok(true, 'sandbox ended')
-})
+1. Inside `./src/events/hit-counter` create a `package.json` file with an empty `dependencies` entry:
```
Good news, we do Lambda treeshaking now! No need to have individual `package.json` files, just add `@architect/functions` to root `package.json` and we'll pick it up at hydration / deploy time.

ryanblock commented on src/views/docs/en/guides/developer-experience/local-development.md:
````diff
@@ -4,112 +4,103 @@ category: Developer experience
+arc sandbox
 ```
+
+Architect will hydrate your [shared code](/docs/en/guides/developer-experience/sharing-code) and run the updated Architect project.
````
We didn't mention adding shared code as part of this local dev guide, so this is a bit awkward. But it'd def make sense to demonstrate shared code.

filmaj replied:
Agreed, perhaps we can reword this to state that _sandbox_ will hydrate function code and link to the `arc hydrate` docs while we are at it?

```suggestion
[`arc sandbox`](../../reference/cli/sandbox) will automatically [hydrate - that is, install needed dependencies](../../reference/cli/hydrate) - for every function making up your Architect project.
```

ryanblock commented on test/sandbox-http-test.js:
```diff
@@ -0,0 +1,21 @@
+  t.plan(4)
+
+  await sandbox.start({ quiet: true })
+  t.ok(true, `sandbox started on ${host}`)
```
`t.ok(true)` → `t.pass()`

filmaj commented on src/views/docs/en/guides/developer-experience/local-development.md:
```diff
@@ -4,112 +4,103 @@ category: Developer experience
-  let result = await tiny.get({ url: 'http://localhost:3333' })
-  t.ok(result, 'got 200 response')
-})
+Your new event may require some dependencies. For this example, [@architect/functions](https://github.com/architect/functions) will be helpful to handle the event subscription. Dependencies can be installed directly to the `hit-counter` directory:
```
Maybe we can link to the [dependency management guide](https://arc.codes/docs/en/guides/developer-experience/dependency-management) in here somehow to note the two approaches to dependency management? 
Perhaps something like:

```suggestion
Your new event may require some dependencies. For this example, [the @architect/functions runtime helper library](../../reference/runtime/node.js) will be helpful to handle the event subscription. [Dependencies in node.js can be managed](dependency-management) either globally from the root `package.json`, or managed via separate `package.json` files in each function's directory. In this example, let's manage them directly in the event's `src/events/hit-counter` directory:
```

filmaj commented on src/views/docs/en/guides/developer-experience/local-development.md:
```diff
@@ -4,112 +4,103 @@ category: Developer experience
-  await sandbox.end()
-  t.ok(true, 'closed')
-})
+3. From the project root start up the sandbox if it is not already running:
```
```suggestion
3. From the project root start up the [sandbox](../../reference/cli/sandbox) if it is not already running:
```

filmaj commented on src/views/docs/en/guides/developer-experience/local-development.md:
```diff
@@ -4,112 +4,103 @@ category: Developer experience
+
+You can run a front-end application from the same Architect project to communicate with the back-end or use a client to interface with HTTP functions.
+
+Of course, the best way to catch bugs is by [testing your Architect project](/docs/en/guides/developer-experience/testing).
```
```suggestion
Of course, the best way to catch bugs is by [testing your Architect project](testing).
```

filmaj commented on src/views/docs/en/guides/developer-experience/testing.md:
```diff
@@ -0,0 +1,100 @@
+---
+title: Testing
+category: Developer experience
+description: How to test with Architect
```
```suggestion
description: How to test your Architect project
```

filmaj commented on src/views/docs/en/guides/developer-experience/testing.md:
```diff
@@ -0,0 +1,100 @@
+
+### `@http` functions
+
+Use `@architect/sandbox` as a module to test `@http` functions with Node:
```
```suggestion
Use `@architect/sandbox` as a module to test [`@http`](../../reference/app.arc/http) functions with Node:
```

filmaj commented on src/views/docs/en/guides/developer-experience/testing.md:
```diff
@@ -0,0 +1,100 @@
+
+### `@events` and `@queues`
+
+Use `@architect/functions` to publish locally to `@events` or `@queues` for testing.
```
```suggestion
Use [`@architect/functions`](../../reference/runtime/node.js) to publish locally to [`@events`](../../reference/app.arc/events) or [`@queues`](../../reference/app.arc/queues) for testing.
```

filmaj commented on src/views/docs/en/guides/developer-experience/testing.md:
```diff
@@ -0,0 +1,100 @@
+
+### `@tables`
+
+Using `@architect/functions` to interact with DynamoDB `@tables` defined locally.
```
```suggestion
Using `@architect/functions` to interact with DynamoDB [`@tables`](../../reference/app.arc/tables) defined locally.
```

filmaj reviewed (changes requested):
Wow, this is so awesome! Thanks for taking this on! I think this is a massive improvement!

I left a few comments, mostly around wording and dropped a few suggestions around linking to other parts of the docs from relevant areas.

Thank you so much for sending this PR!

tbeseda commented:
Awesome feedback @ryanblock and @filmaj 
I'm going to synthesize this today and update here this afternoon.
Thanks for taking a look.

tbeseda commented:
I included suggested edits and decided to use the project's main `package.json` when demonstrating dependency management.

I also added `markdown-it-attrs` in order to add `target="_blank"` to links that leave arc.codes. This will also be useful for adding other attributes while continuing to stick with Markdown instead of too much HTML.  
also added a bit of CSS to show an icon next to external links. I can update all external links in a future PR if maintainers like this idea.

<img width="576" alt="external link" src="https://user-images.githubusercontent.com/15697/128648743-e8f14106-be58-442e-8200-a975ce23015c.png">

Last, Ryan mentioned an issue with the encoding on characters used in the folder structure diagram (see his screenshot above). I wasn't able to reproduce this. I tried with Node 12, 14, and 16. Also fresh installs of the repo after clearing my npm cache.  
It may relate to the occasional refresh issue I've seen when the sandbox restarts and the browser uses locally cached assets. Maybe another contributor can pull and boot up the app to check.

tbeseda commented:
I'm closing this as we're making more targeted passes on docs. I will maintain the branch as a source for improvements to the text of the doc. Some changes here add dependencies and might be out of scope in the near future.
